### PR TITLE
Global discussion re: datagrams (was : 'added C interface to datagram API')

### PR DESCRIPTION
**Pull request by xanathar**
As said in https://github.com/cloudflare/quiche/issues/430 I built a small C interop interface on top of the quic datagram's API (no h3).

This PR mostly to discuss it.

Of note: `quiche_conn_dgram_free` is meant to be called to free the buffer returned by `quiche_conn_dgram_recv` which  was allocated "Rust side". Its `_conn` parameter is totally useless, but I kept it there for API uniformity.

### CODE REVIEW

**Comment by LPardue** [comment]
Thanks for the PR! 

I think there'll be some debate about both the Rust and C APIs, so lets keep this PR open.

**Comment by xanathar** [comment]
> I think there'll be some debate about both the Rust and C APIs, so lets keep this PR open.

Any way I can contribute more to get that discussion going ? I'm happy to help, if possible.

**Comment by LPardue** [comment]
Our main focus right now is in the core protocol, and DATAGRAM is a bit best effort. The good news is that the changes it requires are not too intensive and the API surface seems small. 

As always, having some data to aid decision making is real helpful. Perhaps it is possible to profile the copy vs. non-copy case and understand the tradeoffs a bit more?

**Comment by xanathar** [comment]
I haven't ack'ed this before, but I'm slowly working towards it.

**Comment by xanathar** [comment]
Dumping the benchmarks and thoughts here.

I set up a benchmark which could run in several modes on localhost, that is raw udp (it just floods the client with packets, with no care whether they are received or not), "nocopy" which is the dgram interface which doesn't require an extra data copy (i.e. yours: `dgram_recv(&mut self) -> Result<Vec<u8>>`) or "copy" which is `dgram_recv(&mut self, out: &mut [u8])`.

I did the benchmark in C, given the results (spoiler) I'm not too keen in performing them again in pure Rust, unless there are concerns over this.

The code for the benchmark is very quick and dirty and is at : https://github.com/xanathar/quiche/tree/datagram_iface_perf_comp/performance_comparison . I'm sorry for the dirty-ness of the code but that required a lot of back-and-forth and is written as a throw-away.

To reach a decent amount of bandwidth in a controlled way, I removed flow-control (which actually should be refactored decently and picked, because, if I'm not mistaken the current draft says that datagrams should not be constrained by flow control) and disabled congestion control by passing a dummy congestion control algorithm. On the congestion control topic, I think there are some checks to be done to see if datagrams contribute correctly to congestion control or not (something I plan to check in the future).

I measured the process cpu time (as returned by `clock_gettime(CLOCK_PROCESS_CPUTIME_ID, ...)`) and the bandwidth sustained after a ramp-up. For the 64K datagrams I aimed for a 128MB ramp-up and 4GB target, for the 1KB datagrams test I aimed for a 256MB target after a 8MB ramp-up. Both server and client process were restarted each iteration, to avoid iterations influencing each other and to work around what seems to be a memory leak in the server code, which is something which must be debugged but off topic for this specific context of deciding the API.

The results (showing p50 over 5 tests for each mode):

Test                | UDP | no-copy | copy
--------------------|-----|---------|-------
1K dgrams, CPU sec  | 2.02 | 5.02   | 5.07
1K dgrams, MiB/s    |  492 |  194   |  194 
64K dgrams, CPU sec | 0.52 | 3.85   | 4.01
64K dgrams, MiB/s   | 7325 |  762   |  763 

The tests gave all fairly stable numbers. For the "copy/64KB" scenario I took several more numbers and gathering them gave the result:

Copy/64K       | Average | Std Dev 
--------------|----------|--------
CPU time       |  3.991 | 0.03858  
Bandwidth MB/s |  768.3 | 9.338

Given this, I'd say that the gains for the "no-copy" scenario are negligible and probably not worth pursuing versus a cleaner interface; there is **maybe** a small CPU time gain in having no copy, but really, it's not going to be noticed unless I got the numbers completely wrong.

On the other hand, the performance difference versus a pure UDP scenario is quite big, but probably would disappear once on a "real" network rather than insanely pumping on the loopback.

Given this, I'd recommend going with `dgram_recv(&mut self, out: &mut [u8])`.

**Comment by LPardue** [comment]
Thank you very much for doing this!

I'll need to digest this a bit more, so I'll followup with a more comprehensive response. But again, this is well appreciated.

**Comment by xanathar** [comment]
I found I misunderstood one part of the code around the send function when max_packet_size was taken into consideration and I thus fixed it and rerun the tests. The conclusion doesn't change much. Posting this mostly in case anybody wants to repeat the results. Numbers taken on a c5n.4xlarge EC2 instance.

MEDIAN | UDP | Copy | No-copy
-------|-----|------|----------
1K dgrams, CPU sec	| 0.506169	| 1.3319005	 | 1.3098825
1K dgrams, MiB/s	| 494.058696	| 185.786007	 | 187.3379035
16K dgrams, CPU sec	| 0.962107	| 6.3041795	 | 6.106201
16K dgrams, MiB/s	| 4250.329674	| 525.773377	 | 525.9469895
			
STDDEV | UDP | Copy | No-copy
-------|-----|------|----------
1K dgrams, CPU sec	| 0.003309591 | 0.00648492 | 0.010422114
1K dgrams, MiB/s	| 3.087541535 | 1.071140286 | 1.66477244
16K dgrams, CPU sec	| 0.010671619 | 0.025372608 | 0.029966428
16K dgrams, MiB/s	| 49.00906114 | 2.376173833 | 3.353950162

**Comment by LPardue** [comment]
A bit snowed under here but expect a response over the weekend :) thanks again

**Comment by xanathar** [comment]
Don't worry about it, take your time.
My plan regarding this currently is to go on a series of tasks (below the list) and then stop working on it until most of the material has been reviewed by you when time allows. I don't want to pressure :) but I don't want to block on my side at this time while I have the ball rolling; after all it's a work where 99% of the thing is getting knowledge and 1% is writing the code that does the job.

The tasks I was planning to face are:
- [x] Perform the test in order to unlock a discussion over the interface
- [x] Get some measures around the datagrams
- [x] Suggest an interface (see Issue on the main project)
- [x] Verify the interface is feasible
- [x] Check some specs corner cases
- [ ] Understand the role of congestion control when sending data made up mostly by datagrams

So far only the last one is remaining, after which I plan to stop and resume when you have more time to dedicate to this side activity.

Thanks btw, great library! :)

**Comment by LPardue** [comment]
So taking a fresh look back at my datagram code and comparing it to other read operations in quiche, I kinda dislike the inconsistency of returning a `Vec`. I think I had my reasons at the time but from an API purist view I'm no longer sold.

So I think a refactor of the internal APIs would probably be a good thing, and then the C API proposal becomes a better fit and I think we can avoid the need of a `_free` function..

**Comment by xanathar** [comment]
Thinking about the near future.

Other than some work around retransmission, I hope to get code to a PR-ready state (using proposal 2 as it seems the one which sparked the interest, but I can switch to 1 or do a more vanilla PR and then submit another one for the send queue customizations). 

Unsurprisingly, it's a rather big PR, with multiple commits.

Would you prefer...

- A large PR or several small incremental PRs ?
- The PR to be sent to the cloudflare/quiche repo or yours ?
- Changes to be rebased on top of the latest cloudflare/quiche/master or not and use a merge ?
- Any other requirement ?

**Comment by LPardue** [comment]
I'd like to concentrate work into this fork in order to minimise noise on the main quiche repo. So a rough plan would be to

step 1) sync up my `datagram-frame` branch in this branch to quiche master and refactor out the webtransport stuff to a separate branch. 

step 2) internal refactor to move away from the current `Vec` code. 

That should leave things in a place to add the C API in one PR, and then the additional changes in one or more PRs against this repo.

This lets us review and selectively land iterative changes for datagram here, and then at some checkpoint surface up that aggregate changes to the main quiche repo.

I was planning on doing steps 1 and 2 at the end of this week. 

WDYT?

**Comment by xanathar** [comment]
I think it makes perfect sense, as it would also give me the chance to re-submit the commits I have (many) in small, affordable batches :)

For step 2, if by that you mean the change of signature of `dgram_recv` and not a deeper change, I did it on my branch (https://github.com/xanathar/quiche/commit/0a5911e410955f7d51fca85e0c1da499ebf49bb5#diff-b4aea3e418ccdb71239b96952d9cddb6R2854) so you can copy/paste or I can submit a PR.

**Comment by LPardue** [comment]
@xanathar for fear of getting lost in rebases and force pushes, I've created an alternative branch that rewrites history by taking out webtransport and squashing everything down into a single commit. I've then added a commit to refactor the API and internals, which makes dgram::DatagramQueue simpler and less duplicative. PTAL

https://github.com/LPardue/quiche/tree/datagram-frame-alt

**Comment by xanathar** [comment]
Took a look and it definitely makes sense to me. Starting tomorrow I'll start sending PRs.

Just one thing. On the issue in the main project I suggested 2 proposals to be able to control the queue of outgoing datagrams: one exposing a function to selectively purge unwanted datagrams from the queue using a closure, and the other exposing a customizable queue. The closure one has an easier C interface and a syntax which is more Rust-idiomatic, the other allows for more customization. Any preference for the PRs or shall I just skip that part for now ?

**Comment by LPardue** [comment]
>  On the issue in the main project I suggested 2 proposals to be able to control the queue of outgoing datagrams: one exposing a function to selectively purge unwanted datagrams from the queue using a closure, and the other exposing a customizable queue. The closure one has an easier C interface and a syntax which is more Rust-idiomatic, the other allows for more customization. Any preference for the PRs or shall I just skip that part for now ?

Good point, thanks for reminding me. My 2c: I am attracted to the customizable queue but now realise that this might cause some difficulty around the C API. The more conservative side of me leans to enhancing the existing model (your approach 1) and making progressive improvement steps that mature the underlying DATAGRAM code in quiche. That will empower people to build more advanced applications and provide more input to future design decisions. If customizable queue emerges as a need, then we can decide later whether it warrants a refactor to bring it in. WDYT?

**Comment by xanathar** [comment]
I'm also more for approach 1 - it gives application developers plenty to do with datagrams and has already enough control built in. If the need arise between user, a more complex solution can be designed (likely, due to users' feedback, in an entirely different way which we can't even imagine right now), in the meantime this will do I think. 

Btw, I started sending PRs; apart a couple of ones, most of them are WIPs because I'm waiting for the last one to perform a deeper complete end-to-end test to check for regressions in the rebase/rewrite process.

I'm also sending them... let's say "without flow control" 😅 Feel free to check them when you have time, with no hurry.

**Comment by xanathar** [comment]
All PRs have been sent, except a minor one regarding qlog support which requires me to check with the qlog people (datagrams are not in the format, even if qvis displays them happily) and that one will take a little more time but it's, like, 10 lines.

I've integrated all the changes in a single branch on my side (https://github.com/xanathar/quiche/tree/lpdgalt-pr-integration-test) and everything seems to be working (both MacOS and Windows with different application code).

As such I marked all PRs as ready.

Btw - sorry again for the boringssl bit and the chaos that ensued to refactor that out of the PRs but this whole thing is.. big as you can see and it's becoming hard to manage.

**Comment by LPardue** [comment]
> Btw - sorry again for the boringssl bit and the chaos that ensued to refactor that out of the PRs but this whole thing is.. big as you can see and it's becoming hard to manage.

No problem, these things easily happen and it had no substantial effect on me.

Thanks for making these small PRs, it will keep discussion points focused and hopefully we can land the less contentious points.

**Comment by xanathar** [comment]
Moved this to draft as it's mainly a centralized discussion now :)

**Comment by xanathar** [comment]
I also moved all other pending PRs as Draft, and will unlock as Ready as soon as I address the feedback there (there's a lot of things on the plate for both of us to keep track of)

**Comment by xanathar** [comment]
All (except the examples one) other PR's are unlocked now.

Just as an overall help, this is the DAG of the open PRs as it's becoming hard to follow. Tell me if you prefer a single big one.

```
              +----+
    +---------|  7 |----------+
    |         +----+          |
    |            |            |
    V            V            V
  +----+      +----+        +----+
  | 10 |      | 11 |        | 13 |
  +----+      +----+        +----+
    |            |            |
    |            |            |
    |         +----+          |
    |         | 12 |          |
    |         +----+          |
    |            |            |
    |            V            |
    |         +----+          |
    +-------->| 14 |<---------+
              +----+
```

**Comment by LPardue** [comment]
Thanks! I'll try to review these soon and knock them down 1-by-1
